File: 4382-count-robot-groups/count-robot-groups.cpp
class Solution {
public:
    int countGroups(vector<int>& position, vector<int>& speed, int distance) {

        int n = position.size();

        int groups = 1; // At least 1 group will be made

        int group_speed = speed[n - 1];

        for (int i = n - 2; i >= 0; i--) {

            if (position[i + 1] - position[i] <= distance) {
                continue;
            }
            else if (speed[i] <= group_speed) {
                groups++;
                group_speed = speed[i];
            }
        }

        return groups;
    }
};

File: 4382-count-robot-groups/count-robot-groups.java
import java.util.*;

class Solution {
    public int countGroups(int[] position, int[] speed, int distance) {

        int n = position.length;

        ArrayList<Integer> s = new ArrayList<>();

        for (int i = 0; i < n; i++) {

            if (i > 0 && position[i] - position[i - 1] <= distance) {
                s.remove(s.size() - 1);
            }

            s.add(speed[i]);
        }

        int ans = 0;
        int minSpeed = Integer.MAX_VALUE;

        for (int i = s.size() - 1; i >= 0; i--) {

            if (s.get(i) <= minSpeed) {
                ans++;
                minSpeed = s.get(i);
            }
        }

        return ans;
    }
}